resources/js/mp-history/content.tsx

```diff
@@ -0,0 +1,179 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import ShowMoreLink from 'components/show-more-link';
+import LegacyMatchEvent from 'interfaces/legacy-match-event-json';
+import LegacyMatchGame from 'interfaces/legacy-match-game-json';
+import LegacyMatch from 'interfaces/legacy-match-json';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+import { bottomPageDistance } from 'utils/html';
+import { trans } from 'utils/lang';
+import UserJson from '../interfaces/user-json';
+import Event from './event';
+import Game from './game';
+
+interface Props {
+  currentGameId?: number;
+  events?: LegacyMatchEvent[];
+  hasNext: boolean;
+  hasPrevious: boolean;
+  isAutoloading: boolean;
+  loadingNext: boolean;
+  loadingPrevious: boolean;
+  loadNext: (state: any) => void; // TODO: type is temporary, will be specified better when `main.coffee` is converted to TS
+  loadPrevious: (state: any) => void; // TODO: type is temporary, will be specified better when `main.coffee` is converted to TS
+  match: LegacyMatch;
+  users: Partial<Record<number, UserJson>>;
+}
+
+interface Snapshot {
+  referenceFunc?: () => number;
+  referencePrev?: number;
+  scrollToLastEvent: boolean;
+}
+
+export interface TeamScores {
+  blue: number;
+  red: number;
+}
+
+export default class Content extends React.PureComponent<Props> {
+  scoresCache: Partial<Record<number, TeamScores>> = {};
+
+  componentDidUpdate(prevProps: Readonly<Props>, prevState: Readonly<any>, snapshot?: Snapshot) {
+    if (snapshot?.scrollToLastEvent) {
+      $(window).stop().scrollTo(document.body.scrollHeight, 500);
+    } else if (snapshot?.referenceFunc) {
+      const referenceCurrent = snapshot.referenceFunc();
+      const documentScrollTopCurrent = window.scrollY;
+      const documentScrollTopTarget = documentScrollTopCurrent + referenceCurrent - snapshot.referencePrev!;
+      window.scrollTo(window.scrollX, documentScrollTopTarget);
+    }
+  }
+
+  getSnapshotBeforeUpdate(prevProps: Readonly<Props>, _: Readonly<any>): Snapshot {
+    const snapshot: Snapshot = {
+      scrollToLastEvent: prevProps.isAutoloading && this.props.isAutoloading && bottomPageDistance() < 10,
+    };
+
+    if (!snapshot.scrollToLastEvent) {
+      const hadEvents = prevProps.events != null && prevProps.events.length > 0;
+      const hasEvents = this.props.events != null && this.props.events.length > 0;
+      if (hadEvents && hasEvents) {
```

Comment:
Bit verbose, but this is the only way that I could find that shuts the compiler up about being able to access the first element inside the block...